### DISCUSSION TITLE
Workspaces: plans about N-dimensional / grid layout and possible configuration?

### DISCUSSION THREAD

**Discussion by pirat89** [labwc/labwc#2204] [Q&A]
Hi \o
I am most likely from a minority of users who actually use grid layout for workspaces, but I realized I have big problems with a flat 1-row layout. I am aware about recently merged initial implementation of cosmic-workspaces in labwc and discussd [ext-workspaces protocol](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/40) that's possibly going to be merged during upcoming year. Both protocols count with N-dimensional layout (`coordinates`).

I know that grid layout is not currently planned regarding the scope document and possibly it will not have a priority now, so I made already at least my hackish implementation, where each workspace has `index` an the grid layout is simulated just by virtual grid mapping. In case of interest, the implementation I am using right now for labwc-0.7.4 is here:
* https://github.com/pirat89/labwc/commits/grid-workspace-0.7.4/

My main question is, how do you think the configuration would look like in future to be able to create `X * Y` grid layout? I am asking as  I would like to at least adjust my hackish solution to useat least possibly compatible configuration.

My current implementation adds possibility to specify `rows` and `cols` for `desktops`:
```xml
   <desktops rows="2" cols="3">
     ...
   </desktops>
```

I am aware that this is kind of in conflict with `number` and defined named workspaces, so for now I set rows & cols to have higher priority above other config values when defined. I tried to cover the behaviour in case of various configuration in the [commit msg](https://github.com/pirat89/labwc/commit/a1bc2e9d2865c6c59f4b6e44f175090db3f20a4f).

Note that I would be ok to try to contribute in future to implement this functionality in better way if wanted, however I have close to zero experience with implementation of graphical servers, compositors, window managers, etc. So I believe that my design and implementation would most likely sucks and someone more skilled would need to polish it. So I do not want to just open a PR to take your free time reviewing most likely lame implementation, especially when this is most likely not between priorities of the project.

**Comment by johanmalm** [comment]
Hey and welcome.

I suspect some more sophisticated workspace layouts would be okay to merge… as long as it’s tidy, scalable, doesn’t complicate things too much, etc.

The one area that I think would be very tough to implement well is different workspaces on different outputs.

I’m not sure how much you contributed to open source projects before, but some quick tips for something like this which isn’t trivial….

Try to get a core developer interested and discuss the early proposals before you go too far down one path. It may pay dividend to describe how it will work in some detail before doing too much coding.

Work on the master branch. In fact, the workspace code has changed a bit since the 0.7 branch.

Do some small contributions (code, issues, whatever) so we get to know you and how you operate. It’s easier to merge more complex stuff that way.

**Comment by pirat89** [comment]
hi @johanmalm,

> Try to get a core developer interested and discuss the early proposals before you go too far down one path.

Well, that's one of reasons I opened this topic. My current hackish/naive implementation of the grid-layout is minimalistic
```
 4 files changed, 243 insertions(+), 49 deletions(-)
```

and as I have to migrate to wayland during these weeks and this functionality is crucial fo me, I gave it a shot to create it for myself at least. Do not take me wrong, I do not plan to extend it rapidly and then ask core developers to look at it. I know that there are too many unknowns for me (expected behaviour on tablet? separate workspaces for outputs? etc.). As I am not developing any GUI at all, I believe that anything what I would try to create would need someone else to polish it - or maybe redesign it - so it fits better to possible future plans and do not block future extensions.

I am rather trying to find out whether there are some ideas/plans related to workspace layouts, to see whether there is actually something useful I could do
*   even when it could be just part of the entire required work, providing limited basic functionality that could be extended in future
* or adjust the solution I use to be let's say more compatible when a proper solution is implemented in the project so the transition is then easier.

> It may pay dividend to describe how it will work in some detail before doing too much coding.

Well, I didn't want to go into details when I have not been sure whether such a functionality is wanted, as based on the scope document it seemed to me that it's not between priorities. 

To describe the functionality , I just need to be able to create 2-dimensional layout of workspaces, where I coud `GoToDesktop`|`SendToDesktop`
*  **left** &  **right** - in this case stay in the same row
* **up** & **down** - in this case stay in the same column

with the possibility to specify wrap. Just like with [setlayout](https://openbox.org/dist/tools/setlayout.c) in openbox. Note that in OB  scrolling with mouse transitioned through desktops by rows, when getting to the most border of the grid lead to the transition to the next row, so all workspaces could be visited like that too. I do not need this one functionality, but just know it works like that in OB, so possibly keyword "next" and "prev" could provide this possibility still, which could be still for free I guess. In case of multiple outputs, going to a different workspace would lead to the switch in all displays. I mean, having multiple displays is for me just like having one bigger workspace, In case I want to "freeze" content in one of monitors, I just toggle a window to be omnipresent.

I do not know how labwc it's working with multiple displays. Especially, how e.g. waybar behaves. I haven't tried that yet as I started to play with wayland and labwc (& other apps) setup relatively recently. But it's something what I will experiment in upcoming days.

> Work on the master branch. In fact, the workspace code has changed a bit since the 0.7 branch.

In fact, the original solution is based on master branch - that time commit  f1141c6 (24 Sep) - you can find it in branch [grid-workspace](https://github.com/pirat89/labwc/tree/grid-workspace). I needed to port it to v0.7.4 to be compatible with other packages (wlroots-0.17) on the system. Until the cosmic-workspace initial implementation that happened few days ago, the code is same for both versions of labwc. I rather shared in the description the one for 0.7.4 as that one is practically archived for me, so in case someone read it later, they can see the same stuff I wanted to share. Which will not be the case for up-to-date master for sure.

> Do some small contributions (code, issues, whatever) so we get to know you and how you operate. It’s easier to merge more complex stuff that way.

Well, in case I find out something small that is missing or that seems buggy, I will try to contribute. Right now, I found out _buggy_ titlelayout - at least in version 0.7.4. I cannot get any icon for `desk` in the titlebar. IOW, following configuration for titlebar has visually no effect in comparison to default:
`<layout>menu,desk:iconfiy,max,close</layout>`.  not sure whether it has been fixed in the meanwhile or whether something else is wrong with my setup. But I do not want to spoil this topic by this particular problem. I will create a separate issue/pr for it if I reproduce it with latest commits. 

I would like to keep this discussion related to workspaces really, to hear something about possible plans or ideas. I know that my current solution is minimalistic and naive.

To describe my current solution simply:
* i just added index to each WS
* introduced mapping of 1-dimensional list to 2-dimensional layout
  * replacing `get_prev` / `get_next` functions by get_(up,down,left,right)_index in relation to the index of the current active WS
* loop through all workspaces until the one with calculated index is discovered
  * I do not expect someone would be using high number of WSs to need to optimize it (definitely not me); keeping it stupid
  * in case of bug, that WS with expected index could not be discovered, log it and stay on the current WS

Configuration is extended by the possibility to set `rows` and `cols` attributes for desktops, and also new values `up` and `down` are introduced for `GoToDektop` and `SendToDesktop`. Copying simply current logic for left/right. Details about behaviour/interaction between specified `number` , defined named workspaces, and `cols` and `rows` attributes are covered by the commit msg mentioned in the description. In short
* cols & rows are optional; if used, user can define both or just one of them to get various behaviours
  * e.g. `cols=1 <rows undefined>` result to workspaces created based on `number` and defined named workspaces organized vertically instead horizontally
  * e.g. `cols=2 <rows undefined>` is same like specifying rectangle 2x1 (cols=2 rows=1), etc.. 
* cols & rows specified values have higher priority than number or defined named workspaces, so the grid is always fully filled, and no WS exist outside of rectangle; spefifying 2x3 grid sets fixed 6 workspaces.
* layout has always shape of rectangle - no other "shapes" are accepted; consider 1x1 as smallest rectangle

* reconfiguration is accepted

**Comment by johanmalm** [comment]
Okay, that's great. Shall we leave it to see if any other 'regular contributor' feel like discussing/shaping this.

If not, I'll try to wrap my head around it. Might have to be after the release in 3 weeks.
